core/rewards_manager.go

```diff
@@ -83,65 +95,105 @@ func (rm *RewardManager) GetRewardTransactionByBlockHash(blockHash []byte) *type
+
+	// Write groupId
 	buffer.Write(groupID)
+	// pack fee
+	buffer.Write(common.Uint64ToByte(packFee))
+
 	if len(targetIds) == 0 {
 		return nil, nil, errors.New("GenerateReward targetIds size 0")
 	}
-	for i := 0; i < len(targetIds); i++ {
-		index := targetIds[i]
-		buffer.Write(group.Members[index])
+
+	// Write target indexes
+	for _, idIdx := range targetIds {
+		// Write the mem idx instead
+		buffer.Write(common.UInt16ToByte(uint16(idIdx)))
 	}
+
 	transaction := &types.Transaction{}
 	transaction.Data = blockHash.Bytes()
 	transaction.ExtraData = buffer.Bytes()
-	if len(buffer.Bytes())%common.AddressLength != 0 {
-		return nil, nil, errors.New("GenerateReward ExtraData Size Invalid")
-	}
+
 	transaction.Value = types.NewBigInt(totalValue / uint64(len(targetIds)))
 	transaction.Type = types.TransactionTypeReward
 	transaction.GasPrice = types.NewBigInt(0)
+	transaction.GasLimit = types.NewBigInt(0)
 	transaction.Hash = transaction.GenHash()
 	return &types.Reward{TxHash: transaction.Hash, TargetIds: targetIds, BlockHash: blockHash, GroupID: groupID, TotalValue: totalValue}, transaction, nil
 }
 
 // ParseRewardTransaction parse a bonus transaction and  returns the group id, targetIds, block hash and transcation value
-func (rm *RewardManager) ParseRewardTransaction(transaction *types.Transaction) ([]byte, [][]byte, common.Hash, *types.BigInt, error) {
+func (rm *RewardManager) ParseRewardTransaction(transaction *types.Transaction) (groupId []byte, targets [][]byte, blockHash common.Hash, packFee *big.Int, err error) {
 	reader := bytes.NewReader(transaction.ExtraData)
 	groupID := make([]byte, common.GroupIDLength)
-	addr := make([]byte, common.AddressLength)
-	if n, _ := reader.Read(groupID); n != common.GroupIDLength {
-		return nil, nil, common.Hash{}, nil, errors.New("ParseRewardTransaction Read GroupID Fail")
+	version, e := reader.ReadByte()
+	if e != nil {
+		err = e
+		return
+	}
+	if version != rewardVersion {
+		err = fmt.Errorf("reward version error")
+		return
+	}
+	if _, e := reader.Read(groupID); e != nil {
+		err = fmt.Errorf("read group id error:%v", e)
+		return
+	}
+	pf := make([]byte, 8)
+	if _, e := reader.Read(pf); e != nil {
+		err = fmt.Errorf("read pack fee error:%v", e)
+		return
+	}
+
+	targetIdxs := make([]uint16, 0)
+	idx := make([]byte, 2)
+
+	for n, e := reader.Read(idx); n > 0 && e == nil; n, e = reader.Read(idx) {
+		if e != nil {
+			err = fmt.Errorf("read target idex error: %v", e)
+			return
+		}
+		targetIdxs = append(targetIdxs, common.ByteToUInt16(idx))
+	}
+
+	group := GroupChainImpl.GetGroupByID(groupID)
+	if group == nil {
+		err = fmt.Errorf("group is nil, id=%v", common.ToHex(groupID))
+		return
 	}
 	ids := make([][]byte, 0)
-	for n, _ := reader.Read(addr); n > 0; n, _ = reader.Read(addr) {
-		if n != common.AddressLength {
-			Logger.Debugf("ParseRewardTransaction Addr Size:%d Invalid", n)
-			break
+
+	for _, idx := range targetIdxs {
+		if idx > uint16(len(group.Members)) {
```

Comment:
should it be as below?
for index, idx := range targetIdxs {
    if index >= uint16(len(group.Members)) {